agent/tcs/client/client.go

```diff
@@ -242,6 +263,44 @@ func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetri
 	return requests, nil
 }
 
+// serviceConnectMetricsToPublishMetricRequests loops over all the SC metrics in a
+// task metric to add SC metrics until the message size is within 1 MB.
+// If adding a SC metric to the message exceeds the 1 MB limit, it will be sent in the new message
+func (cs *clientServer) serviceConnectMetricsToPublishMetricRequests(requestMetadata *ecstcs.MetricsMetadata, taskMetric *ecstcs.TaskMetric,
+	messageTaskMetrics []*ecstcs.TaskMetric, requests []*ecstcs.PublishMetricsRequest) (*ecstcs.TaskMetric, []*ecstcs.TaskMetric, []*ecstcs.PublishMetricsRequest) {
+	tempTaskMetric := *taskMetric
+	tempTaskMetric.ServiceConnectMetricsWrapper = tempTaskMetric.ServiceConnectMetricsWrapper[:0]
+
+	for j, serviceConnectMetric := range taskMetric.ServiceConnectMetricsWrapper {
+		tempTaskMetric.ServiceConnectMetricsWrapper = append(tempTaskMetric.ServiceConnectMetricsWrapper, serviceConnectMetric)
+		messageTaskMetrics = append(messageTaskMetrics, &tempTaskMetric)
+		tempMessage, _ := jsonutil.BuildJSON(ecstcs.NewPublishMetricsRequest(requestMetadata, copyTaskMetrics(messageTaskMetrics)))
```

Comment:
Hmm, I worry about the CPU & memory consumption here since `BuildJSON` will create a lot of garbage and takes processing time. Let's not do premature optimization, but keep an eye on this. We will need to load test & profile this to see if it's an issue or not.  (maybe add a TODO for that here)

Alternatively, JSON has a somewhat predictable overhead. We could make calculations based on the size of the go structs and then add JSON overhead as a percentage.

Reply:
Agreed, will add a TODO for this.